tls/s2n_client_hello.c

```diff
@@ -69,8 +69,11 @@ static S2N_RESULT s2n_generate_client_session_id(struct s2n_connection *conn)
         return S2N_RESULT_OK;
     }
 
-    /* Only generate the session id for TLS1.3 if in middlebox compatibility mode */
-    if (conn->client_protocol_version >= S2N_TLS13 && !s2n_is_middlebox_compat_enabled(conn)) {
+    /* Only generate the session id for TLS1.3 if in middlebox compatibility mode
+     *
+     * actual_protocol_version is used here so the session id is generated if TLS1.3
+     * was not negotiated, even if the client supports it */
```

Comment:
It's probably worth noting [why](https://github.com/aws/s2n-tls/pull/3845/files#r1113730266) we have to use actual_protocol_version here. "was not negotiated" doesn't make a lot of sense on its own since negotiation hasn't happened yet :)

Reply:
> It's probably worth noting [why](https://github.com/aws/s2n-tls/pull/3845/files#r1113730266) we have to use actual_protocol_version here.

Would you say [that code](https://github.com/aws/s2n-tls/blob/67f36f6d313e278ce004570b91cb8cefb3efcee5/tls/s2n_resume.c#L219) is incorrect though? My understanding is the `actual_protocol_version` is the most up to date protocol version we are using given the outcome of any negotiation, and `client_protocol_version` is just the highest protocol we could support, so it makes sense to me that we would general use `actual_protocol_version`. I can add more context in the comment though

> "was not negotiated" doesn't make a lot of sense on its own since negotiation hasn't happened yet :)

In the case of a session resumption, hasn't it happened already and thats why we have this issue?

Reply:
Yes, setting only actual_protocol_version and not client_protocol_version is correct. However, when we refer to "negotiation", we're referring to the process of the client and server negotiating parameters. That HASN'T happened for this connection yet. The server might not even accept our attempt at session resumption. We're just pretending it has happened because weird TLS1.2 session resumption rules.

Reply:
Thanks, I've updated the comment